### ISSUE TITLE
emscripten_fetch returns id 0 if not used with EMSCRIPTEN_FETCH_REPLACE

### ISSUE THREAD

**Issue by slowriot** [emscripten-core/emscripten#23124]
Tested with version 3.1.73.

When calling `emscripten_fetch` with any combination of attributes, but **without** using `EMSCRIPTEN_FETCH_REPLACE`, the `emscripten_fetch_t` struct it returns always has the `id` member set to `0`.  With `EMSCRIPTEN_FETCH_REPLACE`, these increment normally.

Callbacks receive the correct `id` in all cases, just the struct initially returned by `emscripten_fetch` is wrong.

This is a relatively new issue as a recent update has broken previously working code with this, but unfortunately I'm not sure what the last working version was.

Minimal example to reproduce:

```cpp
  emscripten_fetch_attr_t attr;
  emscripten_fetch_attr_init(&attr);
  std::strcpy(attr.requestMethod, "GET");

  attr.attributes = EMSCRIPTEN_FETCH_LOAD_TO_MEMORY | EMSCRIPTEN_FETCH_REPLACE;
  for(unsigned int i{0}; i != 10; ++i} {
    emscripten_fetch_t *fetch{emscripten_fetch(&attr, "example.com")};
    std::cout << "id: " << fetch->id << std::endl;  // incrementing normally: 1, 2, 3, ...
  }

  attr.attributes = EMSCRIPTEN_FETCH_LOAD_TO_MEMORY;
  for(unsigned int i{0}; i != 10; ++i} {
    emscripten_fetch_t *fetch{emscripten_fetch(&attr, "example.com")};
    std::cout << "id: " << fetch->id << std::endl;  // always 0
  }
```

**Comment by slowriot** [comment]
Looking at the logic flow in [Fetch.js](https://github.com/emscripten-core/emscripten/blob/d5c419f8aa428bc3d1d071d23227dd13a00b463c/src/Fetch.js#L560), we have: https://github.com/emscripten-core/emscripten/blob/d5c419f8aa428bc3d1d071d23227dd13a00b463c/src/Fetch.js#L559-L562

In the case that `EMSCRIPTEN_FETCH_REPLACE` is set, we go straight to `fetchXHR`, which then sets `id`: https://github.com/emscripten-core/emscripten/blob/d5c419f8aa428bc3d1d071d23227dd13a00b463c/src/Fetch.js#L308

The trouble is that if `EMSCRIPTEN_FETCH_REPLACE` is not set, and we call [`fetchLoadCachedData`](https://github.com/emscripten-core/emscripten/blob/d5c419f8aa428bc3d1d071d23227dd13a00b463c/src/Fetch.js#L128), then we end up in the getRequest.onsuccess callback in the second branch: https://github.com/emscripten-core/emscripten/blob/d5c419f8aa428bc3d1d071d23227dd13a00b463c/src/Fetch.js#L165-L174

Only at this point do we call the `onerror` callback, which calls either `performCachedXhr` or `performUncachedXhr`.  But this happens async, so manwhile the `fetchLoadCachedData` function has returned without setting `id` to any value.

I think it might make more sense, if there's a possibility that an xhr will be made, to allocate an id anyway with `Fetch.xhrs.allocate`, so this can be returned immediately.

**Comment by slowriot** [comment]
Note if I set `-sFETCH_SUPPORT_INDEXEDDB=0`, we don't see this problem, as we go straight to the XHRs without trying to query the db async, so we always non-zero `id`s in that case.

**Comment by sbc100** [comment]
Yes, it looks like `id` tracking is only for outstanding XHR request, not from IDB requests.  I don't know why that is.

I'm also not sure why we need an ID at all since the address of the `emscripten_fetch_t` would make a perfectly good unique id I think.   I guess its too late to change that.

**Comment by sbc100** [comment]
BTW, as the user of the API, why do you care about the ID that is stored in the struct?   Don't all the callbakc functions take a unique `emscripten_fetch_t` pointer?

**Comment by yejun81** [comment]
and according to this,  call emscripten_fetch_close would lead to memory leak
https://github.com/emscripten-core/emscripten/blob/b94d6e6ced33576a1ad96c7f6e701788c6f49211/system/lib/fetch/emscripten_fetch.c#L161

**Comment by slowriot** [comment]
> and according to this, call emscripten_fetch_close would lead to memory leak
> 
> https://github.com/emscripten-core/emscripten/blob/b94d6e6ced33576a1ad96c7f6e701788c6f49211/system/lib/fetch/emscripten_fetch.c#L161

No, once you're in a callback, the id has been populated.  You wouldn't call `emscripten_fetch_close` before one of the associated callbacks has been called.  But it's a pain that the id isn't consistent.

**Comment by yejun81** [comment]
> > and according to this, call emscripten_fetch_close would lead to memory leak
> > https://github.com/emscripten-core/emscripten/blob/b94d6e6ced33576a1ad96c7f6e701788c6f49211/system/lib/fetch/emscripten_fetch.c#L161
> 
> No, once you're in a callback, the id has been populated. You wouldn't call `emscripten_fetch_close` before one of the associated callbacks has been called. But it's a pain that the id isn't consistent.

No.when you fetch data with EMSCRIPTEN_FETCH_NO_DOWNLOAD  ,the id would never populate.

**Comment by slowriot** [comment]
> > > and according to this, call emscripten_fetch_close would lead to memory leak
> > > https://github.com/emscripten-core/emscripten/blob/b94d6e6ced33576a1ad96c7f6e701788c6f49211/system/lib/fetch/emscripten_fetch.c#L161
> > 
> > 
> > No, once you're in a callback, the id has been populated. You wouldn't call `emscripten_fetch_close` before one of the associated callbacks has been called. But it's a pain that the id isn't consistent.
> 
> No.when you fetch data with EMSCRIPTEN_FETCH_NO_DOWNLOAD ,the id would never populate.

Oh that's interesting, I hadn't considered that path.  It sounds like the ID assignment process does need to be reconsidered, for sure.

**Comment by yejun81** [comment]
> > > > and according to this, call emscripten_fetch_close would lead to memory leak
> > > > https://github.com/emscripten-core/emscripten/blob/b94d6e6ced33576a1ad96c7f6e701788c6f49211/system/lib/fetch/emscripten_fetch.c#L161
> > > 
> > > 
> > > No, once you're in a callback, the id has been populated. You wouldn't call `emscripten_fetch_close` before one of the associated callbacks has been called. But it's a pain that the id isn't consistent.
> > 
> > 
> > No.when you fetch data with EMSCRIPTEN_FETCH_NO_DOWNLOAD ,the id would never populate.
> 
> Oh that's interesting, I hadn't considered that path. It sounds like the ID assignment process does need to be reconsidered, for sure.

yes, like sbc100 has mentioned, it look likes id tracking is only for outstanding XHR request, not from IDB requests.when use emscripten_fetch_close to manage IDB request, memory leak appeared.

**Comment by Myrindd** [comment]
I am a developer making heavy use of IDB caching for downloads in a 3D engine, and this basically systematically crashes my system after hogging all the memory. My temporary fix is to create a custom version of the emscripten_fetch_close which ignore "fetch->id == 0" so all the memory is always freed.

However, I have to recreate the emscripten_fetch_free function. Is the below function a correct implementation?

void emscripten_fetch_free(uint32_t fetch_id)
{
    EM_ASM ({
        __emscripten_fetch_free($0);
    }, fetch_id);
}

**Comment by yejun81** [comment]
> I am a developer making heavy use of IDB caching for downloads in a 3D engine, and this basically systematically crashes my system after hogging all the memory. My temporary fix is to create a custom version of the emscripten_fetch_close which ignore "fetch->id == 0" so all the memory is always freed.
> 
> However, I have to recreate the emscripten_fetch_free function. Is the below function a correct implementation?
> 
> void emscripten_fetch_free(uint32_t fetch_id) { EM_ASM ({ __emscripten_fetch_free($0); }, fetch_id); }

i am a developer working on 3D Engine and heavy use of IDB caching for downloads too,O(∩_∩)O.
why you have to recreate emscripten_fetch_free?

**Comment by Myrindd** [comment]
I guess we are in the same club then! :)
_emscripten_fetch_free() is part of emscripten_internal.h; which does not appear possible to include in projects.

**Comment by yejun81** [comment]
so we are not alone.^_^
since _emscripten_fetch_free just do some work with xhr request, so i think it doesn't need to recreate it.

**Comment by AlexandreK38** [comment]
@slowriot Hi! Has the id been or will be patched? I just faced the same issue on the 4.0.11 and was using the id to track the requests but I would need to switch to the fetch address for my map

**Comment by sbc100** [comment]
I don't believe this was fixed yet no.
